packages/react-router-dev/vite/route-chunks.ts

```diff
@@ -266,55 +276,108 @@ function getDependentIdentifiersForPath(
     identifiers: new Set(),
   };
 
+  // Ensure we don't recurse indefinitely
+  if (visited.has(path)) {
+    return identifiers;
+  }
+
   visited.add(path);
 
   // Recursively traverse the AST to find all identifiers the path depends on.
   path.traverse({
     Identifier(path) {
+      // We can skip all of this work if we've already processed this identifier.
+      if (identifiers.has(path)) {
+        return;
+      }
+
       identifiers.add(path);
+
       let binding = path.scope.getBinding(path.node.name);
-      if (binding?.path && !visited.has(binding.path)) {
-        getDependentIdentifiersForPath(binding.path, { visited, identifiers });
+
+      if (!binding) {
+        return;
+      }
+
+      getDependentIdentifiersForPath(binding.path, { visited, identifiers });
+
+      // Trace all references to the identifier
+      for (let reference of binding.referencePaths) {
+        // Each export declaration is handled separately in our chunking logic
+        // so we don't want to trace the entire export statement, otherwise all
+        // identifiers in the export statement will be marked as dependencies
+        // and we won't be able to split chunks sharing this export statement.
+        if (reference.isExportNamedDeclaration()) {
+          continue;
+        }
+
+        getDependentIdentifiersForPath(reference, {
+          visited,
+          identifiers,
+        });
+      }
+
+      // For completeness we also want to trace constant violations since, even
+      // though the code results in a runtime error, it still compiles.
+      for (let constantViolation of binding.constantViolations) {
+        getDependentIdentifiersForPath(constantViolation, {
+          visited,
+          identifiers,
+        });
       }
     },
   });
 
-  // If the identifier is part of a destructuring assignment, we include all
-  // other identifiers in the expression as dependencies.
+  let topLevelStatement = getTopLevelStatementPathForPath(path);
+  let withinImportStatement = topLevelStatement.isImportDeclaration();
+  let withinExportStatement = topLevelStatement.isExportDeclaration();
+
+  // Include all identifiers in the top-level statement as dependencies, except
+  // for import/export statements since they have more fine-grained filtering.
+  if (!withinImportStatement && !withinExportStatement) {
+    getDependentIdentifiersForPath(topLevelStatement, {
+      visited,
+      identifiers,
+    });
+  }
+
+  // Destructuring assignments in export statements have more fine-grained
+  // filtering, so we include all identifiers in the expression as dependencies.
   if (
+    withinExportStatement &&
```

Comment:
This code was already targeting export statements specifically, it just wasn't explicitly scoped in the code. Now that we have much more general tracing of dependencies above, we can scope this extra check much more aggressively.